File: clients/go/input.go
package sharptown

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strings"
)

// Input is an image source for a transform. It is consumed lazily at request time and is
// stream-oriented: nothing is read until the transform runs, and the transports stream
// straight from it — so an image can be processed entirely in memory or from a network
// stream, without ever touching disk.
//
// Use File or Reader to pass a standard file object or any io.Reader; use Bytes for an
// in-memory buffer; Path and URL are conveniences that read from disk or fetch over HTTP.
type Input struct {
	filename    string
	contentType string
	open        func(ctx context.Context, hc *http.Client) (io.ReadCloser, error)
}

// Reader wraps any io.Reader (e.g. an *os.File, a multipart.File, an http body, a
// bytes.Buffer). The reader is not closed by the client.
func Reader(r io.Reader, filename string) Input {
	name := orDefault(filename, "image")
	return Input{
		filename:    name,
		contentType: guessContentType(name),
		open: func(context.Context, *http.Client) (io.ReadCloser, error) {
			return io.NopCloser(r), nil
		},
	}
}

// File wraps a standard *os.File, taking its base name as the filename. The file is not
// closed by the client.
func File(f *os.File) Input {
	return Reader(f, filepath.Base(f.Name()))
}

// Bytes wraps an in-memory image buffer.
func Bytes(data []byte, filename string) Input {
	name := orDefault(filename, "image")
	return Input{
		filename:    name,
		contentType: guessContentType(name),
		open: func(context.Context, *http.Client) (io.ReadCloser, error) {
			return io.NopCloser(bytes.NewReader(data)), nil
		},
	}
}

// Path reads the image from a file path when the transform runs.
func Path(p string) Input {
	name := filepath.Base(p)
	return Input{
		filename:    name,
		contentType: guessContentType(name),
		open: func(context.Context, *http.Client) (io.ReadCloser, error) {
			return os.Open(p)
		},
	}
}

// URL fetches the image over HTTP when the transform runs.
func URL(rawURL string) Input {
	name := "image"
	if parsed, err := url.Parse(rawURL); err == nil {
		if base := path.Base(parsed.Path); base != "" && base != "/" && base != "." {
			name = base
		}
	}
	return Input{
		filename:    name,
		contentType: guessContentType(name),
		open: func(ctx context.Context, hc *http.Client) (io.ReadCloser, error) {
			if hc == nil {
				hc = http.DefaultClient
			}
			req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
			if err != nil {
				return nil, &Error{Message: err.Error()}
			}
			resp, err := hc.Do(req)
			if err != nil {
				return nil, &Error{Message: fmt.Sprintf("failed to fetch input from %s: %s", rawURL, err)}
			}
			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
				resp.Body.Close()
				return nil, &Error{Message: fmt.Sprintf("failed to fetch input from %s: %d", rawURL, resp.StatusCode), Status: resp.StatusCode}
			}
			return resp.Body, nil
		},
	}
}

func orDefault(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}

func guessContentType(filename string) string {
	switch strings.ToLower(strings.TrimPrefix(filepath.Ext(filename), ".")) {
	case "jpg", "jpeg":
		return "image/jpeg"
	case "png":
		return "image/png"
	case "webp":
		return "image/webp"
	case "gif":
		return "image/gif"
	case "avif":
		return "image/avif"
	case "heif", "heic":
		return "image/heif"
	case "tif", "tiff":
		return "image/tiff"
	case "bmp":
		return "image/bmp"
	default:
		return "application/octet-stream"
	}
}

File: clients/flutter/lib/src/input.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import 'error.dart';

/// A resolved image: its raw bytes plus the filename and content type used in the request.
class ResolvedInput {
  /// Creates a resolved input.
  const ResolvedInput(this.bytes, this.filename, this.contentType);

  /// The raw image bytes.
  final Uint8List bytes;

  /// The filename sent in the multipart request.
  final String filename;

  /// The MIME type guessed from the filename.
  final String contentType;
}

/// An image source for a transform. It is resolved lazily, when the transform runs.
///
/// Use [ImageInput.bytes] for an in-memory buffer, [ImageInput.file] for a `dart:io`
/// [File], [ImageInput.path] to read from disk, or [ImageInput.url] to fetch over HTTP.
///
/// ```dart
/// ImageInput.bytes(buffer, 'photo.jpg');
/// ImageInput.file(File('photo.jpg'));
/// ImageInput.path('photo.jpg');
/// ImageInput.url('https://example.com/cat.jpg');
/// ```
class ImageInput {
  const ImageInput._(this._resolve, this._filename);

  final Future<ResolvedInput> Function(http.Client httpClient) _resolve;
  final String _filename;

  /// Wraps an in-memory image buffer.
  factory ImageInput.bytes(Uint8List data, [String filename = 'image']) {
    return ImageInput._(
      (_) async => ResolvedInput(data, filename, _guessContentType(filename)),
      filename,
    );
  }

  /// Wraps a `dart:io` [File], taking its base name as the filename.
  factory ImageInput.file(File file) => ImageInput.path(file.path);

  /// Reads the image from a file path when the transform runs.
  factory ImageInput.path(String path) {
    final name = _basename(path);
    return ImageInput._(
      (_) async {
        final file = File(path);
        if (!file.existsSync()) {
          throw SharptownError('Input file not found: $path');
        }
        return ResolvedInput(
            await file.readAsBytes(), name, _guessContentType(name));
      },
      name,
    );
  }

  /// Fetches the image over HTTP when the transform runs.
  factory ImageInput.url(String url) {
    var name = 'image';
    final parsed = Uri.tryParse(url);
    if (parsed != null && parsed.pathSegments.isNotEmpty) {
      final last = parsed.pathSegments.last;
      if (last.isNotEmpty) name = last;
    }
    return ImageInput._(
      (httpClient) async {
        final http.Response response;
        try {
          response = await httpClient.get(Uri.parse(url));
        } catch (error) {
          throw SharptownError('Failed to fetch input from $url: $error');
        }
        if (response.statusCode < 200 || response.statusCode >= 300) {
          throw SharptownError(
            'Failed to fetch input from $url: ${response.statusCode}',
            status: response.statusCode,
          );
        }
        return ResolvedInput(response.bodyBytes, name, _guessContentType(name));
      },
      name,
    );
  }

  /// The filename this input will report in the request.
  String get filename => _filename;

  /// Resolves the input to bytes, using [httpClient] for [ImageInput.url] sources.
  Future<ResolvedInput> resolve(http.Client httpClient) => _resolve(httpClient);

  static String _basename(String path) {
    final normalized = path.replaceAll('\\', '/');
    final segment = normalized.split('/').last;
    return segment.isEmpty ? 'image' : segment;
  }

  static String _guessContentType(String filename) {
    final dot = filename.lastIndexOf('.');
    final ext = dot == -1 ? '' : filename.substring(dot + 1).toLowerCase();
    switch (ext) {
      case 'jpg':
      case 'jpeg':
        return 'image/jpeg';
      case 'png':
        return 'image/png';
      case 'webp':
        return 'image/webp';
      case 'gif':
        return 'image/gif';
      case 'avif':
        return 'image/avif';
      case 'heif':
      case 'heic':
        return 'image/heif';
      case 'tif':
      case 'tiff':
        return 'image/tiff';
      case 'bmp':
        return 'image/bmp';
      default:
        return 'application/octet-stream';
    }
  }
}
